src/single_random_draw.rs

```diff
@@ -0,0 +1,432 @@
+        input.witness.iter().map(|elem| elem.len()),
+    );
+
+    let input_weight = transaction::predict_weight(Some(vec![weight_predict]), None::<Vec<_>>);
+    let input_fee: Option<Amount> = fee_rate.checked_mul_by_weight(input_weight);
+
+    match input_fee {
+        Some(f) => Ok(utxo.value.checked_sub(f)),
+        None => Err(Error::Multiplication(input_weight, fee_rate)),
+    }
+}
+
+/// Randomly select coins for the given target by shuffling the utxo pool and
+/// taking UTXOs until the given target is reached.
+///
+/// ## Parameters
+/// ///
+/// /// * `target` - target value to send to recipient.
+/// /// * `fee_rate` - ratio of transaction amount per size.
+/// /// * `utxos` - UTXOs from which to sum the target amount.
+/// /// * `inputs` - the order corresponds to the ordering of the UTXOs where the first input consumes the first UTXO and so on.
+/// /// * `transaction` - the transaction to populate with resulting inputs and outputs.
+/// /// * `script_pubkey` - needed to compose the transaction outputs.
+/// /// * `rng` - used primarily by tests to make the selection deterministic.
+///
+pub fn select_coins_srd<R: rand::Rng + ?Sized>(
+    target: Amount,
+    fee_rate: FeeRate,
+    utxos: &mut [TxOut],
+    inputs: &mut [TxIn],
+    transaction: &mut Transaction,
+    script_pubkey: ScriptBuf,
+    rng: &mut R,
+) -> Result<Option<()>, Error> {
+    utxos.shuffle(rng);
+
+    let zipped = utxos.iter().zip(inputs.iter());
+    let mut transaction_inputs: Vec<TxIn> = Vec::new();
+
+    let mut value = Amount::ZERO;
+    for (utxo, input) in zipped {
+        let effective_value: Option<Amount> = get_effective_value(utxo, input, fee_rate)?;
+
+        // skip if effective_value is negative
+        match effective_value {
+            Some(e) => value += e,
+            None => continue,
+        }
+
+        transaction_inputs.push(input.clone());
+
+        if value >= target + CHANGE_LOWER {
+            // value is >= than target so the subtraction will not be negative.
+            let change_amount: Amount = value - target;
+
+            let change_output: TxOut =
+                TxOut { value: change_amount, script_pubkey: script_pubkey.clone() };
+
+            let destination_output: TxOut =
+                TxOut { value: target, script_pubkey: script_pubkey.clone() };
+
+            transaction.input = transaction_inputs.clone();
+            transaction.output = vec![change_output, destination_output];
+
+            let weight: Weight = transaction.weight();
+            let transaction_fee: Option<Amount> = fee_rate.checked_mul_by_weight(weight);
+
+            match transaction_fee {
+                Some(f) => {
+                    // if the fee is very high then change_amount - fee could be negative.
+                    // This will therefore be None in such a case so that more inputs are
+                    // selected.
+                    let effective_value = change_amount.checked_sub(f);
```

Comment:
Another way to handle this is by including the fee for the change output in the target for the change budget, see e.g. https://github.com/bitcoin/bitcoin/pull/27846

Reply:
I meant that if the “change budget” is supposed to cover fees, but it is possible for the change budget to be smaller than the fee here and you need to go back to select more inputs, then you should probably set the change budget such that it already explicitly includes the fees.

Reply:
About the paramater [change_fee](https://github.com/murchandamus/bitcoin/blob/1771daa815ec014276cfcb30c934b0eaff4d72bf/src/wallet/test/fuzz/coinselection.cpp#L68) that gets passed in to SRD; What I think you are suggesting is just add the amount for creating a change_output and add that to the budget.  so  `if value >= target + CHANGE_LOWER {` becomes `if value >= target + CHANGE_LOWER + change_fee`.  However that doesn't preclude the possibility for `let effective_value = change_amount.checked_sub(f);` to be negative if fees are extreme?

Reply:
I created a draft commit https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/23/commits/d6d4d4d7373cb02424f8a1e0d9f8773c38f86f79 to try out your suggestion.  However there is a failing test now since adding change_fee to the budget doesn't preclude the fee from being larger than the change output.

Reply:
I'm going to undo the draft commit since it's obviously broken.  I'll take a look more into cores implementation since there is a detail I think I'm missing.  Another item on the short list will be to add fuzz testing.  Unless I hear otherwise I'll mark this conversation resolved.

Reply:
> Another way to handle this is by including the fee for the change output in the target for the change budget

I made some minor change to calculate the fee for the outputs and transaction as part of the change budget.